jscomp/bsb_helper/bsb_helper_packer.ml

```diff
@@ -22,46 +22,85 @@
  * along with this program; if not, write to the Free Software
  * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. *)
 
+#if BS_NATIVE then
 type pack_t = PackBytecode | PackNative
 
-let pack pack_byte_or_native ~batch_files ~includes =
-  let suffix_object_files, suffix_library_files, compiler = begin match pack_byte_or_native with
-  | PackBytecode -> Literals.suffix_cmo, Literals.suffix_cma , "ocamlc.opt"
-  | PackNative   -> Literals.suffix_cmx, Literals.suffix_cmxa, "ocamlopt.opt"
+let ( // ) = Ext_path.combine
+
+(* The packer is called with object files (.cmo / .cmx) which will be namespaced and we're using
+   those names to read-in the mlast files which are not namespaced. So we strip the namespace
+   before reading them in. *)
+let module_of_filename filename = 
+  let str = Ext_filename.chop_extension_maybe filename in
+  match (String.rindex str '-') with 
+  | exception Not_found -> str
+  | len -> String.sub str 0 len
+
+let pack pack_byte_or_native ~batch_files ~includes ~namespace ~warnings ~warn_error ~verbose ~cwd =
+  let suffix_object_files, suffix_library_files, compiler, nested = begin match pack_byte_or_native with
+  | PackBytecode -> Literals.suffix_cmo, Literals.suffix_cma , "ocamlc", "bytecode"
+  | PackNative   -> Literals.suffix_cmx, Literals.suffix_cmxa, "ocamlopt", "native"
   end in
   let module_to_filepath = Ext_list.fold_left batch_files  String_map.empty
     (fun m v ->
       String_map.add m
-      (Ext_filename.module_name v)
+      (Ext_filename.module_name (module_of_filename v))
       (Ext_filename.chop_extension_maybe v)
       )  
   in
   let dependency_graph = Ext_list.fold_left batch_files String_map.empty
     (fun m file ->
+      let module_name = module_of_filename file in
+      let suffix = if Sys.file_exists (module_name ^ Literals.suffix_mlast) then Literals.suffix_mlast
+        else Literals.suffix_reast in
       String_map.add m
-        (Ext_filename.module_name file)
-        (Bsb_helper_extract.read_dependency_graph_from_mlast_file ((Ext_filename.chop_extension_maybe file) ^ Literals.suffix_mlast))
+        (Ext_filename.module_name module_name)
+        (Bsb_helper_extract.read_dependency_graph_from_mlast_file (module_name ^ suffix))
         )    
   in
   let domain =
     String_map.fold dependency_graph String_set.empty 
       (fun k _ acc -> String_set.add acc k)
       in
   let sorted_tasks = Bsb_helper_dep_graph.sort_files_by_dependencies ~domain dependency_graph in
-  let list_of_object_files = Queue.fold
+  let all_object_files = Queue.fold
     (fun acc v -> match String_map.find_opt module_to_filepath v with
       | Some file -> (file ^ suffix_object_files) :: acc
       | None -> failwith @@ "build.ninja is missing the file '" ^ v ^ "' that was used in the project. Try force-regenerating but this shouldn't happen."
       )
     []
     sorted_tasks in
+  let warning_command = if String.length warnings > 0 then
```

Comment:
note to help me review it easier, the changes to bsb_native alone could be sent separately and merged quickly